internal/app/query/query_service.go

```diff
@@ -325,3 +371,146 @@ func (tc *ticketCache) update() {
 	logger.Debugf("Ticket Cache update: Previous %d, Deleted %d, Fetched %d, Current %d", previousCount, deletedCount, len(toFetch), len(tc.tickets))
 	tc.err = nil
 }
+
+type backfillCache struct {
```

Comment:
Simply copy pasting here created a lot of duplicated code, which I mostly don't like because the common part is tricky concurrency stuff.

I think the best solution here is to refactor the ticketCache into simply `cache`.  Include in the struct (and `newCache`) an update function with type `func(store statestore.Service, state interface{}) error`, moving both update functions to static functions which can be passed into the constructor.  You'll need a default value for the state as well.  

Then change f in request to be  ` f func(interface{})`.  For the update and request functions/closures, simply do a cast.

It's not pretty, and generics would be really nice here.  However all of the casts will trivially be tested by the e2e tests, so I'm not really worried about a bug there.  I think the cost of a few casts is worth de-duplicating the concurrency code.
Thoughts?

Reply:
I also saw lots of duplicate code here, just thought it might be united into one set of functions later on at the refactoring stage. In order to leave beta (backfill) and release (ticket) code separate.
Overall there is only one condition which is changed, it is comparing Generation number in a memory (map) and in HSET in Redis.

Reply:
Some further examples to clarify:

```golang
type ticketCache struct {
	store statestore.Service

	requests chan *cacheRequest

	// Single item buffered channel.  Holds a value when runQuery can be safely
	// started.  Basically a channel/select friendly mutex around runQuery
	// running.
	startRunRequest chan struct{}

	wg sync.WaitGroup

        update func(statestore.Service, interface{}) error

	// Mutlithreaded unsafe fields, only to be written by update, and read when
	// request given the ok.
        value interface{}
	err     error
}
```


```golang
func (s *queryService) QueryTicketIds(req *pb.QueryTicketIdsRequest, responseServer pb.QueryService_QueryTicketIdsServer) error {
       /// <code removed here for example simplicity>
	var results []string
	err = s.tc.request(ctx, func(value interface{}) {
                tickets := value.(map[string]*pb.Ticket)
		for id, ticket := range tickets {
			if pf.In(ticket) {
				results = append(results, id)
			}
		}
	})
	if err != nil {
		err = errors.Wrap(err, "QueryTicketIds: failed to run request")
		return err
	}
       /// <code removed here for example simplicity>
}
```

```golang
func updateTicketCache(store statestore.Service, value interface{}) error {
  tickets := value.(map[string]*pb.Ticket)
   /// etc....
}
```

Reply:
Covered by this PR #1300 . We can close this PR I think.